website/upgrade-guides/0-13.html.markdown

```diff
@@ -0,0 +1,367 @@
+
+In order to establish the hierarchical namespace, Terraform now requires
+explicit source information for any providers that are not HashiCorp-maintained,
+using a new syntax in the `required_providers` nested block inside the
+`terraform` configuration block:
+
+```hcl
+terraform {
+  required_providers {
+    azurerm = {
+      # The "hashicorp" namespace is the new home for the HashiCorp-maintained
+      # provider plugins.
+      #
+      # source is not required for the hashicorp/* namespace as a measure of
+      # backward compatibility for commonly-used providers, but recommended for
+      # explicitness.
+      source  = "hashicorp/azurerm"
+      version = "~> 2.12"
+    }
+    datadog = {
+      # source is required for providers in other namespaces, to avoid ambiguity.
+      source  = "terraform-providers/datadog"
+      version = "~> 2.7.0"
+    }
+  }
+}
+```
+
+If you are using providers that now require an explicit source location to be
+specified, `terraform init` will produce an error like the following:
+
+```
+Error: Failed to install providers
+
+Could not find required providers, but found possible alternatives:
+
+  hashicorp/datadog -> terraform-providers/datadog
+  hashicorp/fastly -> terraform-providers/fastly
+
+If these suggestions look correct, upgrade your configuration with the
+following command:
+    terraform 0.13upgrade
+```
+
+As mentioned in the error message, Terraform v0.13 includes an automatic
+upgrade command `terraform 0.13upgrade` that is able to automatically generate
+source addresses for unlabelled providers by consulting the same lookup table
+that was previously used for Terraform v0.12 provider installation. This command
+will automatically modify the configuration of your current module, so you can
+use the features of your version control system to inspect the proposed changes
+before committing them.
+
+We recommend running `terraform 0.13upgrade` even if you don't see the message,
+because it will generate the recommended explicit source addresses for
+providers in the "hashicorp" namespace.
+
+For more information on declaring provider dependencies, see
+[Provider Requirements](/docs/configuration/provider-requirements.html).
+That page also includes some guidance on how to write provider dependencies
+for a module that must remain compatible with both Terraform v0.12 and
+Terraform v0.13; the `terraform 0.13upgrade` result includes a conservative
+version constraint for Terraform v0.13 or later, which you can weaken to
+`>= 0.12.26` if you follow the guidelines in
+[v0.12-Compatible Provider Requirements](/docs/configuration/provider-requirements.html#v012-compatible-provider-requirements).
+
+After you've added explicit provider source addresses to your configuration,
+run `terraform init` again to re-run the provider installer.
+
+-> **Action:** Either run `terraform 0.13upgrade` for your module, or manually update the provider declarations to use explicit source addresses.
```

Comment:
Somewhere in here we could mention that *all* modules need to be 0.13upgraded (or otherwise modified to include provider source), since terraform evaluates providers required by each module individually. That's already come up with a couple of users (possibly also a good place to mention the `terraform providers` command).

I know we don't want to add too many words to each section, so perhaps we consider adding a "common issues" / "trouble shooting" section at the end?

Reply:
I've changed this to say "for _each of your modules_", and also added a paragraph above explaining that this is a per-module migration.